### PR TITLE
Quantum Signal Processing time-evolution

### PR DESCRIPTION
**Pull request by JamesB-1qbit**
The Quantum Signal Processing algorithm for Hamiltonian Simulation as defined by arXiv:2002.11649.

- Uses pyqsp for the phases by default.
- More stable/faster function that uses interface oct2py with octave based QSPPACK also available

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -29,5 +29,5 @@ def install(package):
     packages=setuptools.find_packages(),
     test_suite="tangelo",
     setup_requires=['h5py'],
-    install_requires=['h5py', 'bitarray', 'openfermion', 'openfermionpyscf']
+    install_requires=['h5py', 'bitarray', 'openfermion', 'openfermionpyscf', 'pyqsp']
```
Comment (ValentinS4t1qbit) [inline setup.py:32]: I wonder if we should require this in the install, considering its just a single feature, maintained by a third party. These other requirements here are more important or closer to python core.

Maybe we should have a try / error block that tells the user "this feature requires pyqsp, please install the pyqsp python package".

File: tangelo/toolboxes/circuits/lcu.py [hunk 2]
```diff
@@ -20,6 +20,7 @@
 
 import math
 from typing import Union, Tuple, List
+from copy import copy
```
Comment (ValentinS4t1qbit) [inline tangelo/toolboxes/circuits/lcu.py:23]: To make a shallow copy of a list just use the built-in copy method or slicing.
If your control lists contain integers (e.g immutable values) then the shallow copy is equivalent to a deepcopy.

```
l = [1,2,3]
l1 = l[:] # Full copy, via slicing
l2 = l.copy() # Shallow copy. Ok with integers etc

l.append('a')
l1.append('b')
l2.append('c')

# Each list is indeed separate / different
print(l, l1, l2)
```

File: tangelo/toolboxes/circuits/qsp.py [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+        coef[i-1][0] = (-1)**(i-1)*jn(2*(i-1), tau)
+
+    coef[0] = coef[0]/2
+    phi1, _ = octave.QSP_solver(coef, 0, opts, nout=2)
+
+    coef = np.zeros((maxorder//2+1, 1), dtype=float, order='F')
+    for i in range(1, len(coef)+1):
+        coef[i-1][0] = (-1)**(i-1)*jn(2*i-1, tau)
+    phi2, _ = octave.QSP_solver(coef, 1, opts, nout=2)
+    return list(phi1.flatten()), list(phi2.flatten())
```
Comment (ValentinS4t1qbit) [inline tangelo/toolboxes/circuits/qsp.py:127]: skip line before L127

File: tangelo/toolboxes/circuits/qsp.py [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+    octave.addpath(folder)
+    opts = octave.struct("criteria", eps)
+    maxorder = math.ceil(1.4*tau+np.log(1e14))
+    if maxorder % 2 == 1:
+        maxorder -= 1
+
+    coef = np.zeros((maxorder//2 + 1, 1), dtype=float, order='F')
+    for i in range(1, len(coef)+1):
+        coef[i-1][0] = (-1)**(i-1)*jn(2*(i-1), tau)
+
```
Comment (ValentinS4t1qbit) [inline tangelo/toolboxes/circuits/qsp.py:119]: Remove empty line to mimic the other similar code block

File: tangelo/toolboxes/circuits/qsp.py [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+    if control is not None:
+        c_list += control if isinstance(control, list) else [control]
+
+    anc = m_qs[-1]+1
+    if with_z:
+        qubcirc = Circuit([Gate("CH", anc, control=c_list), Gate("CZ", anc, control=c_list)])
+    else:
+        qubcirc = Circuit([Gate("CH", anc, control=c_list)])
+
+    circ_angles = []
```
Comment (ValentinS4t1qbit) [inline tangelo/toolboxes/circuits/qsp.py:169]: L170 defeats the point of L169

File: tangelo/toolboxes/circuits/qsp.py [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+        Circuit: The QSP Hamiltonian simulation circuit with oblivious amplitude amplification to ensure very high success probability.
+    """
+
+    if control is not None:
+        control_list = copy(control) if isinstance(control, list) else [control]
+    else:
+        control_list = []
+
+    # If tau is negative, flip sign of tau and qu_op.
+    flip_tau_sign = False
```
Comment (ValentinS4t1qbit) [inline tangelo/toolboxes/circuits/qsp.py:220]: set that variable to `flip_tau_sign = (tau < 0.)`.
Then I think L221-224 can be replaced by a duplicate of L268-270.

File: tangelo/toolboxes/circuits/qsp.py [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+    circ += get_qsp_circuit_no_anc(cua, m_qs, anglesr, control=lcu_qs+control_list, with_z=False)
+    circ += Circuit([Gate("X", lcu_qs[0])])
+    # imaginary part i*sin(Ht)
+    circ += get_qsp_circuit_no_anc(cua, m_qs, anglesi, control=lcu_qs+control_list, with_z=False)
+    # -I to ensure probability is np.arcsin
+    circ += Circuit([Gate("X", lcu_qs[1]), Gate("CRZ", target=0, parameter=2*np.pi, control=lcu_qs+control_list), Gate("X", lcu_qs[1])])
+
+    circ += uprep.inverse()
+
+    if control is not None:
```
Comment (ValentinS4t1qbit) [inline tangelo/toolboxes/circuits/qsp.py:263]: Out of curiosity: do we need the branching ?
What if we only use the first branch, and control=None is passed ?

It seems to make sense to me that control=None should naturally mean it's not a controlled gate and would return something no different than the non-controlled version of the gate, as expected.

File: tangelo/toolboxes/circuits/qsp.py [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+"""Module to generate the circuits necessary to implement quantum signal processing (QSP) Hamiltonian Simulation
+
+Ref:
+[1]: Yulong Dong, Xiang Meng, K. Birgitta Whaley, Lin Lin, "Efficient phase-factor evaluation in quantum signal processing"
+2021, arXiv:2002.11649 (https://arxiv.org/abs/2002.11649)
+"""
+
+import math
+from typing import Union, List, Tuple
+from copy import copy
```
Comment (ValentinS4t1qbit) [inline tangelo/toolboxes/circuits/qsp.py:24]: I dont think you need that (see other comments)

File: tangelo/toolboxes/circuits/qsp.py [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+    Args:
+        qubit_list (List[int]): The qubits controlled by zero.
+        target (int): The target qubit.
+        control (List[int] or int): The qubits controlled by one.
+
+    Returns:
+        Circuit: The zero controlled cnot circuit with possible extra controls on 1.
+    """
+    x_ladder = [Gate("X", q) for q in qubit_list]
+    gates = x_ladder + [Gate("CX", target=target, control=qubit_list+control)] + x_ladder
```
Comment (ValentinS4t1qbit) [inline tangelo/toolboxes/circuits/qsp.py:142]: Control can be an int or a union of int. What if control is an int here? Does this work ?

**Review by ValentinS4t1qbit — CHANGES_REQUESTED** [review]
A few things that could be improved or cleaned up, but nothing big

File: tangelo/toolboxes/circuits/qsp.py (hunk 7 above)
Comment (JamesB-1qbit) [reply tangelo/toolboxes/circuits/qsp.py:265]: Yes, in general that is correct. Our current implementation forces users to specify controlled gates with a `C` in front of the gate name.  We can certainly revisit that if you want.

Comment (ValentinS4t1qbit) [reply tangelo/toolboxes/circuits/qsp.py:265]: We can revisit that later if a user deems it a concern.

**Review by ValentinS4t1qbit — APPROVED** [review]
Thank you james !

**Pull request merged**